Packages/io.chainsafe.web3-unity/Runtime/Scripts/Connection/ConnectionHandler.cs

```diff
@@ -1,88 +1,68 @@
-        public ConnectionProvider ConnectionProvider { get; private set; }
-
-        private void Start()
-        {
-            Initialize();
-        }
 
+        public ConnectionProvider[] Providers => providers;
+        
         /// <summary>
         /// Initializes Connection Handler.
         /// </summary>
-        protected virtual async void Initialize()
+        public async Task Initialize()
         {
             Web3BuilderServiceAdapters = GetComponents<IWeb3BuilderServiceAdapter>();
 
-            Web3InitializedHandlers = GetComponents<IWeb3InitializedHandler>();
-            
-            foreach (var provider in providers)
+            foreach (var provider in Providers)
             {
-                if (provider != null && provider.IsAvailable)
-                {
-                    var instantiatedProvider = connectModal.AddProvider(provider);
-
-                    await instantiatedProvider.Initialize();
-                    
-                    instantiatedProvider.ConnectButton.onClick.AddListener(delegate
-                    {
-                        ConnectionProvider = instantiatedProvider;
-                        
-                        ConnectClicked();
-                    });
-                }
+                await provider.Initialize();
             }
         }
 
-        private async void ConnectClicked()
+        public async Task<CWeb3> Restore()
         {
-            await TryConnect();
+            var data = new StoredConnectionProviderData();
+
+            await data.LoadOneTime();
+
+            var provider = Providers.OfType<RestorableConnectionProvider>()
+                .SingleOrDefault(p => p.GetType() == data.Type);
+            
+            if (provider != null && provider.RememberSession && await provider.SavedSessionAvailable())
+            {
+                return await (this as IConnectionHandler).Connect(provider);
+            }
+
+            return null;
         }
-        
-        /// <summary>
-        /// Try to Connect and displays error and throws exception on a failed attempt.
-        /// </summary>
-        public virtual async Task TryConnect()
+
+        public Web3Builder ConfigureServices(Web3Builder web3Builder)
         {
-            try
+            return web3Builder.Configure(services =>
             {
-                connectModal.ShowLoading();
+                var web3InitializedHandlers = GetComponents<IWeb3InitializedHandler>();
 
-                await (this as IConnectionHandler).Connect();
-            }
-            catch (Exception e)
-            {
-                if (!(e is TaskCanceledException))
+                foreach (var handler in web3InitializedHandlers)
                 {
-                    connectModal.DisplayError(
-                        $"Connection failed, please try again.");
-                    
-                    ConnectionProvider.HandleException(e);
+                    services.AddSingleton(handler);
```

Comment:
Woudn't this cause an issue? since you're not specifying the type, but it will always replace the type?

Reply:
nope doesn't happen because they're different concrete types, if that was the case we would only have one `ILifeCycleParticipant`

Reply:
checked to make sure as well